crates/sui-core/src/event_handler.rs

```diff
@@ -1,78 +1,127 @@
-        Streamer::spawn(rx);
+        let streamer = Streamer::spawn(EVENT_DISPATCH_BUFFER_SIZE);
         Self {
             module_cache: SyncModuleCache::new(ResolverWrapper(validator_store)),
-            streamer_queue: tx,
+            event_streamer: streamer,
         }
     }
 
     pub async fn process_events(&self, effects: &TransactionEffects) {
-        // serializely dispatch event processing to honor events' orders.
+        // serially dispatch event processing to honor events' orders.
         for event in &effects.events {
             if let Err(e) = self.process_event(event).await {
-                error!(error =? e, "Failed to send EventEnvolope to dispatch");
+                error!(error =? e, "Failed to send EventEnvelope to dispatch");
             }
         }
     }
 
     pub async fn process_event(&self, event: &Event) -> SuiResult {
-        let envolope = match event {
-            Event::MoveEvent { .. } => {
+        let json_value = match event {
+            Event::MoveEvent(event_obj) => {
                 debug!(event =? event, "Process MoveEvent.");
-                match event.extract_move_struct(&self.module_cache) {
-                    Ok(Some(move_struct)) => {
-                        let json_value = serde_json::to_value(&move_struct).map_err(|e| {
-                            SuiError::ObjectSerializationError {
-                                error: e.to_string(),
-                            }
-                        })?;
-                        EventEnvelope::new(get_unixtime_ms(), None, event.clone(), Some(json_value))
+                let move_struct = event_obj.to_move_struct_with_resolver(
+                    ObjectFormatOptions::default(),
+                    &self.module_cache,
+                )?;
+                Some(serde_json::to_value(&move_struct).map_err(|e| {
+                    SuiError::ObjectSerializationError {
+                        error: e.to_string(),
                     }
-                    Ok(None) => unreachable!("Expect a MoveStruct from a MoveEvent."),
-                    Err(e) => return Err(e),
-                }
+                })?)
             }
-            _ => EventEnvelope::new(get_unixtime_ms(), None, event.clone(), None),
+            _ => None,
         };
-
+        let envelope = EventEnvelope::new(get_unix_time_ms(), None, event.clone(), json_value);
         // TODO store events here
+        self.event_streamer.send(envelope).await
+    }
+
+    pub fn subscribe(&self, filter: EventFilter) -> impl Stream<Item = EventEnvelope> {
+        self.event_streamer.subscribe(filter)
+    }
+}
+
+#[derive(Eq, PartialEq, Ord, PartialOrd, Clone, Debug, Serialize, Deserialize, JsonSchema)]
+pub struct EventFieldFilter {
+    #[serde(default)]
+    pub fields: BTreeMap<String, String>,
+}
+
+#[derive(Eq, PartialEq, Ord, PartialOrd, Clone, Debug, Serialize, Deserialize, JsonSchema)]
+pub enum EventFilter {
```

Comment:
Discussed with @patrickkuo but here are my thoughts:
1. EventFilter should be for native EventEnvelope type filtering, and the RPC layer should convert arguments into native types for faster and more correct processing;
2. EventFilter should start with simpler, per-field filters, and more complex ones can be built on top of the simple ones.
3. We should try to re-use these filter definitions for both Event storage and subscriptions

For example here is something that meets the goals above:

    pub enum EventFilter {
        ByObjectID(ObjectID),
        ByPackageID(ObjectID),
        ByEventType(EventType),
        ByMoveField(EventFieldFilter),
        ByModuleName(String),
        ByContract { packageID: ObjectID, module_name: Option<String>, event_filters: EventFieldFilter },
    }

Reply:
Thus ByContract would use the filtering of ByPackageID, ByModuleName, and byMoveField together.

Reply:
This is great suggestion, I will make that changes